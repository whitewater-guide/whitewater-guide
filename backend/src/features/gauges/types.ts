import { loadGraphqlFile } from '../../apollo/loadGraphqlFile';
import { RawTimestamped } from '../../db';
import { NamedResource } from '../../apollo/types';

export const GaugesSchema = loadGraphqlFile('gauges');

export interface GaugeRaw extends NamedResource, RawTimestamped {
  code: string;
  source_id: string;
  request_params?: any;
  cron?: string | null;
  last_timestamp?: Date | null;
  level_unit?: string | null;
  flow_unit?: string | null;
  last_level?: number | null;
  last_flow?: number | null;
  url?: string | null;
  enabled?: boolean;
}
