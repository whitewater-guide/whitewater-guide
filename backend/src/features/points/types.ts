/**
 * Raw row from database `points` table
 */
export interface PointRaw {
  id: string;
  name: string | null;
  description: string | null;
  kind: string;
  coordinates: string;
}
