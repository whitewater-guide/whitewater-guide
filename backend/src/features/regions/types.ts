import { ArrayMaxSize, IsDefined, IsInt, IsUUID, Length, Max, Min } from 'class-validator';
import { Geometry, Point, Polygon } from 'wkx';
import { NamedResource } from '../../apollo';
import { loadGraphqlFile } from '../../apollo/loadGraphqlFile';
import { RawTimestamped } from '../../db';
import { Timestamped } from '../../db/types';

export const RegionsSchema = loadGraphqlFile('regions');

/**
 * Raw row from database `source` table
 */
export interface RegionRaw extends NamedResource, RawTimestamped {
  description: string | null;
  season: string | null;
  season_numeric: number[];
  bounds: string | null;
  hidden: boolean | null;
}

/**
 * This is graphql type
 */
export interface Region extends NamedResource, Timestamped {
  description: string | null;
  season: string | null;
  seasonNumeric: number[];
  bounds: Array<[number, number, number]> | null;
  hidden: boolean | null;
}

export class RegionInput {
  @IsUUID()
  id?: string;

  @IsDefined()
  @Length(3, 128)
  name: string;

  description?: string;

  season?: string;

  @IsDefined()
  @IsInt({ each: true })
  @Min(0, { each: true })
  @Max(23, { each: true })
  @ArrayMaxSize(24)
  seasonNumeric?: number[];

  // TODO: add validation
  bounds?: Array<[number, number, number]>;

  hidden?: boolean;
}

export function toRaw(input: RegionInput): Partial<RegionRaw> {
  const { seasonNumeric, bounds, ...rest } = input;
  let rawBounds = null;
  if (bounds) {
    const polygon: Polygon = new Polygon(bounds.map(p => new Point(...p)));
    // Close the ring
    polygon.exteriorRing.push(new Point(...bounds[0]));
    polygon.srid = 4326;
    rawBounds = polygon.toEwkt();
  }
  return {
    ...rest,
    season_numeric: seasonNumeric ? seasonNumeric : [],
    bounds: rawBounds,
  };
}
