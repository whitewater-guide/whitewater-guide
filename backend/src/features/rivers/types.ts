import { loadGraphqlFile } from '../../apollo/loadGraphqlFile';
import { RawTimestamped } from '../../db';
import { NamedNode } from '../../ww-commons';
import { RegionRaw } from '../regions';

export const RiversSchema = loadGraphqlFile('rivers');

export interface RiverRaw extends NamedNode, RawTimestamped {
  region_id: string;
  region?: RegionRaw;
  alt_names: string[];
}
