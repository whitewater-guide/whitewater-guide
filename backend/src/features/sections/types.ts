import { loadGraphqlFile } from '../../apollo/loadGraphqlFile';
import { RawTimestamped } from '../../db';
import { NamedNode, Tag } from '../../ww-commons';
import { PointRaw } from '../points/types';

export const SectionsSchema = loadGraphqlFile('sections');

export interface GaugeBindingRaw {
  minimum?: number;
  maximum?: number;
  optimum?: number;
  impossible?: number;
  approximate?: boolean;
}

export interface SectionRaw extends NamedNode, RawTimestamped {
  river_id: string;
  gauge_id: string | null;
  season_numeric: number[];
  levels: GaugeBindingRaw | null;
  flows: GaugeBindingRaw | null;
  shape: string;
  distance: number | null;
  drop: number | null;
  duration: number | null;
  difficulty: number;
  difficulty_xtra: string | null;
  rating: number | null;
  description: string | null;
  season: string | null;
  flows_text: string | null;
  put_in: string;
  take_out: string;

  pois: PointRaw[] | null;
  tags: Tag[] | null;

  count: number | null; // window function count
}

export interface SectionsFilter {
  riverId?: string;
  regionId?: string;
  updatedAfter?: Date;
}
