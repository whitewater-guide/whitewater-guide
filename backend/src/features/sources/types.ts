import { IsDefined, IsEnum, IsUrl, IsUUID, Length } from 'class-validator';
import { NamedResource } from '../../apollo';
import { loadGraphqlFile } from '../../apollo/loadGraphqlFile';
import { RawTimestamped } from '../../db';
import { Timestamped } from '../../db/types';
import { isCron } from '../../util/isCron';

export const SourcesSchema = loadGraphqlFile('sources');

export enum HarvestMode {
  ALL_AT_ONCE = 'allAtOnce',
  ONE_BY_ONE = 'oneByOne',
}

/**
 * Raw row from database `source` table
 */
export interface SourceRaw extends NamedResource, RawTimestamped {
  terms_of_use: string | null;
  script: string;
  cron: string | null;
  harvest_mode: HarvestMode;
  url: string | null;
  enabled: boolean | null;
}

/**
 * This is graphql type
 */
export interface Source extends NamedResource, Timestamped {
  termsOfUse: string | null;
  script: string;
  cron: string | null;
  harvestMode: HarvestMode;
  url: string | null;
  enabled: boolean | null;
}

export class SourceInput {
  @IsUUID()
  id?: string;

  @Length(3, 128)
  name: string;

  termsOfUse?: string;

  @IsDefined()
  script: string;

  @isCron()
  cron?: string;

  @IsEnum(HarvestMode)
  harvestMode: HarvestMode;

  @IsUrl()
  url?: string;
}

export function inputToRaw(input: SourceInput): Partial<SourceRaw> {
  return {
    id: input.id,
    name: input.name,
    terms_of_use: input.termsOfUse,
    script: input.script,
    cron: input.cron,
    harvest_mode: input.harvestMode,
    url: input.url,
  };
}
