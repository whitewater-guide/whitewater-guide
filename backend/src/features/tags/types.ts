import { loadGraphqlFile } from '../../apollo/loadGraphqlFile';

export const TagsSchema = loadGraphqlFile('tags');
