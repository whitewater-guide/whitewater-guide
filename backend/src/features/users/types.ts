import * as passport from 'passport';
import { NamedResource } from '../../apollo';
import { loadGraphqlFile } from '../../apollo/loadGraphqlFile';
import { RawTimestamped, Timestamped } from '../../db';

export const UsersSchema = loadGraphqlFile('users');

export enum Role {
  USER = 1,
  ADMIN = 2,
  SUPERADMIN = 3,
}

export interface UserRaw extends NamedResource, RawTimestamped {
  avatar: string | null;
  email: string | null;
  role: number;
}

/**
 * This is graphql type
 */
export interface User extends NamedResource, Timestamped {
  avatar: string | null;
  email: string | null;
  role: number;
}

export interface LoginRaw extends RawTimestamped {
  user_id: string;
  id: string;
  provider: string;
  username: string;
  tokens: {[key: string]: string};
  profile: passport.Profile;
}
