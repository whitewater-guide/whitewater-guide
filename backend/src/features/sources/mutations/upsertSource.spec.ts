import db, { holdTransaction, rollbackTransaction } from '../../../db';
import { adminContext, anonContext, userContext } from '../../../test/context';
import { isTimestamp, isUUID, noTimestamps, noUnstable, runQuery } from '../../../test/db-helpers';
import { HarvestMode, SourceInput } from '../../../ww-commons';
import { SourceRaw } from '../types';

beforeEach(holdTransaction);
afterEach(rollbackTransaction);

const requiredSource: SourceInput = {
  id: null,
  name: 'New Source',
  harvestMode: HarvestMode.ALL_AT_ONCE,
  script: 'newScript',
  url: null,
  termsOfUse: null,
  cron: null,
  enabled: null,
};

const optionalSource: SourceInput = {
  id: '6d0d717e-aa9d-11e7-abc4-cec278b6b50a',
  name: 'Updated source',
  script: 'updatedScript',
  cron: '1 1 * * *',
  harvestMode: HarvestMode.ONE_BY_ONE,
  url: 'http://google.com',
  termsOfUse: 'New terms of use',
  enabled: null,
};

const mutation = `
  mutation upsertSource($source: SourceInput!, $language: String){
    upsertSource(source: $source, language: $language){
      id
      language
      name
      termsOfUse
      script
      cron
      harvestMode
      url
      enabled
      createdAt
      updatedAt
    }
  }
`;

describe('resolvers chain', () => {
  test('anon should not pass', async () => {
    const result = await runQuery(mutation, { source: requiredSource }, anonContext);
    expect(result.errors).toBeDefined();
    expect(result.data).toBeDefined();
    expect(result.data!.upsertSource).toBeNull();
    expect(result).toMatchSnapshot();
  });

  test('user should not pass', async () => {
    const result = await runQuery(mutation, { source: requiredSource }, userContext);
    expect(result.errors).toBeDefined();
    expect(result.data).toBeDefined();
    expect(result.data!.upsertSource).toBeNull();
    expect(result).toMatchSnapshot();
  });

  test('should throw on invalid input', async () => {
    const input: SourceInput = {
      id: 'invalid-input',
      name: 'Invalid source',
      script: 'updatedScript',
      cron: '300 1 * * *',
      harvestMode: HarvestMode.ONE_BY_ONE,
      url: 'not url',
      termsOfUse: 'New terms of use',
      enabled: null,
    };
    const result = await runQuery(mutation, { source: input }, adminContext);
    expect(result.errors).toBeDefined();
    expect(result.data).toBeDefined();
    expect(result.data!.upsertSource).toBeNull();
    expect(result).toMatchSnapshot();
  });
});

describe('insert', () => {
  let insertResult: any;
  let insertedSource: any;

  beforeEach(async () => {
    insertResult = await runQuery(mutation, { source: requiredSource }, adminContext);
    insertedSource = insertResult && insertResult.data && insertResult.data.upsertSource;
  });

  afterEach(() => {
    insertResult = null;
    insertedSource = null;
  });

  test('should return result', async () => {
    expect(insertResult.errors).toBeUndefined();
    expect(insertResult.data).toBeDefined();
    expect(insertResult.data!.upsertSource).toBeDefined();
  });

  test('should add one more source', async () => {
    const result = await db().table('sources').count();
    expect(result[0].count).toBe('3');
  });

  test('should have id', () => {
    expect(insertedSource.id).toBeDefined();
    expect(isUUID(insertedSource.id)).toBe(true);
  });

  test('should have timestamps', () => {
    expect(insertedSource.createdAt).toBeDefined();
    expect(insertedSource.updatedAt).toBeDefined();
    expect(isTimestamp(insertedSource.createdAt)).toBe(true);
    expect(isTimestamp(insertedSource.updatedAt)).toBe(true);
  });

  test('should match snapshot', () => {
    expect(noUnstable(insertedSource)).toMatchSnapshot();
  });

});

describe('update', () => {
  let oldSource: SourceRaw | null;
  let updateResult: any;
  let updatedSource: any;

  beforeEach(async () => {
    oldSource = await db().table('sources').where({ id: optionalSource.id }).first();
    updateResult = await runQuery(mutation, { source: optionalSource }, adminContext);
    updatedSource = updateResult && updateResult.data && updateResult.data.upsertSource;
  });

  afterEach(() => {
    updateResult = null;
    updatedSource = null;
    oldSource = null;
  });

  test('should return result', async () => {
    expect(updateResult.errors).toBeUndefined();
    expect(updateResult.data).toBeDefined();
    expect(updateResult.data!.upsertSource).toBeDefined();
  });

  test('should not change total number of sources', async () => {
    const result = await db().table('sources').count();
    expect(result[0].count).toBe('2');
  });

  test('should have id', () => {
    expect(updatedSource.id).toBe(optionalSource.id);
  });

  test('should update updated_at timestamp', () => {
    expect(updatedSource.createdAt).toBe(oldSource!.created_at.toISOString());
    expect(new Date(updatedSource.updatedAt).valueOf()).toBeGreaterThan(oldSource!.updated_at.valueOf());
  });

  test('should match snapshot', () => {
    expect(noTimestamps(updatedSource)).toMatchSnapshot();
  });

});

describe('i18n', () => {
  const norwayRu = {
    id: '786251d4-aa9d-11e7-abc4-cec278b6b50a',
    script: 'norway',
    name: 'Норвегия',
    termsOfUse: 'Правила пользования Норвегией',
    cron: '0 * * * *',
    harvestMode: HarvestMode.ONE_BY_ONE,
    url: 'http://nve.no',
    enabled: false,
  };

  const galiciaRu = {
    id: '6d0d717e-aa9d-11e7-abc4-cec278b6b50a',
    script: 'galicia',
    cron: '0 * * * *',
    harvestMode: HarvestMode.ALL_AT_ONCE,
    url: 'http://ya.ru',
    enabled: false,
    name: 'Новая Галисия',
    termsOfUse: 'Правила пользования новой галисией',
  };

  it('should add new translation', async () => {
    const result = await runQuery(mutation, { source: norwayRu, language: 'ru' }, adminContext);
    expect(result.data!.upsertSource).toMatchObject({
      id: '786251d4-aa9d-11e7-abc4-cec278b6b50a',
      name: 'Норвегия',
      language: 'ru',
    });
    const translation = await db().table('sources_translations').select()
      .where({ source_id: '786251d4-aa9d-11e7-abc4-cec278b6b50a', language: 'ru' })
      .first();
    expect(translation.name).toBe('Норвегия');
  });

  it('should modify common props in other language', async () => {
    await runQuery(mutation, { source: norwayRu, language: 'ru' }, adminContext);
    const commons = await db().table('sources').select('harvest_mode')
      .where({ id: '786251d4-aa9d-11e7-abc4-cec278b6b50a' }).first();
    expect(commons.harvest_mode).toBe(HarvestMode.ONE_BY_ONE);
  });

  it('should modify translation', async () => {
    const result = await runQuery(mutation, { source: galiciaRu, language: 'ru' }, adminContext);
    expect(result.errors).toBeUndefined();
    expect(result.data!.upsertSource).toMatchObject({
      id: '6d0d717e-aa9d-11e7-abc4-cec278b6b50a',
      language: 'ru',
      name: 'Новая Галисия',
      termsOfUse: 'Правила пользования новой галисией',
    });
    const translation = await db().table('sources_translations').select()
      .where({ source_id: '6d0d717e-aa9d-11e7-abc4-cec278b6b50a', language: 'ru' })
      .first();
    expect(translation).toMatchObject({
      source_id: '6d0d717e-aa9d-11e7-abc4-cec278b6b50a',
      language: 'ru',
      name: 'Новая Галисия',
      terms_of_use: 'Правила пользования новой галисией',
    });
  });
});
