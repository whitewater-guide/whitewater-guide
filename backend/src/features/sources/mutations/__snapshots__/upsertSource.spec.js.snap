// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`insert should match snapshot 1`] = `
Object {
  "cron": null,
  "harvestMode": "allAtOnce",
  "name": "New Source",
  "script": "newScript",
  "termsOfUse": null,
  "url": null,
}
`;

exports[`update should match snapshot 1`] = `
Object {
  "id": "b28f728a-6efa-11e7-96fe-47a2910087ec",
  "cron": "1 1 * * *",
  "harvestMode": "oneByOne",
  "name": "Updated source",
  "script": "updatedScript",
  "termsOfUse": "New terms of use",
  "url": "http://google.com",
}
`;
